### PR TITLE
ENH: set method for AxisLabels in the ctkWidgetAxes

### PR DESCRIPTION
**Pull request by Punzo**
The aim is to enable the customization of the ctkAxesWidget in Slicer.

### CODE REVIEW

File: Libs/Widgets/ctkAxesWidget.cpp [hunk 1]
```diff
@@ -293,6 +293,13 @@ void ctkAxesWidget::paintEvent(QPaintEvent *)
 }
 
 // ----------------------------------------------------------------------------------
+void ctkAxesWidget::setAxesLabels(QStringList Labels)
```
Comment (jcfr) [inline Libs/Widgets/ctkAxesWidget.cpp]: `QStringList Labels` -> `const QStringList& labels`

File: Libs/Widgets/ctkAxesWidget.cpp [hunk 2]
```diff
@@ -293,6 +293,13 @@ void ctkAxesWidget::paintEvent(QPaintEvent *)
...
+{
+  Q_D(ctkAxesWidget);
```
Comment (jcfr) [inline Libs/Widgets/ctkAxesWidget.cpp]: print a warning if number of labels is different from `6` and return `false`

**Comment by jcfr** [comment]
Thanks for the patch :+1:

**Comment by jcfr** [comment]
Beside of few nitpicks, it should soon be good for integration.

Please, consider also updating the test `/Libs/Widgets/Testing/Cpp/ctkAxesWidgetTest1.cpp`

File: Libs/Widgets/Testing/Cpp/ctkAxesWidgetTest1.cpp [hunk 3]
```diff
@@ -84,6 +85,19 @@ int ctkAxesWidgetTest1(int argc, char * argv [] )
...
+  axesLabels <<  "W";
+  axesLabels <<  "E";
+  axesLabels <<  "S";
+  axesLabels <<  "N";
+  axesLabels <<  "Z";
+  axesLabels <<  "z";
+  if (!axes.setAxesLabels(AxesLabels))
+    {
+    return EXIT_FAILURE; 
+    }
```
Comment (jcfr) [inline Libs/Widgets/Testing/Cpp/ctkAxesWidgetTest1.cpp]: Check also that axes.axesLabels() return the expected list.

You could include the new `ctkCoreTestingMacros.h` and use `CHECK_QSTRINGLIST()`

File: Libs/Widgets/ctkAxesWidget.cpp [hunk 4]
```diff
@@ -293,6 +293,21 @@ void ctkAxesWidget::paintEvent(QPaintEvent *)
 }
 
 // ----------------------------------------------------------------------------------
+bool ctkAxesWidget::setAxesLabels(const QStringList& Labels)
```
Comment (jcfr) [inline Libs/Widgets/ctkAxesWidget.cpp]: `Labels` -> `labels`

File: Libs/Widgets/ctkAxesWidget.cpp [hunk 5]
```diff
@@ -293,6 +293,21 @@ void ctkAxesWidget::paintEvent(QPaintEvent *)
...
+{
+  Q_D(ctkAxesWidget);
+  if (Labels.size() != 6)
+    {
+    qWarning("ctkAxesWidget::setAxesLabels() "
+             "failed: the input StringList is not of dimension 6.");
```
Comment (jcfr) [inline Libs/Widgets/ctkAxesWidget.cpp]: -> `ctkAxesWidget::setAxesLabels() failed: Exactly 6 labels are expected.`

**Comment by jcfr** [comment]
Consider also adding a Q_PROPERTY similar to what is done here: https://github.com/Punzo/CTK/blob/b9d3d92391359338e2a6c45b45e4968c9aeedc58/Libs/Widgets/ctkAxesWidget.h#L39

File: Libs/Widgets/Testing/Cpp/ctkAxesWidgetTest1.cpp [hunk 6]
```diff
@@ -84,6 +86,17 @@ int ctkAxesWidgetTest1(int argc, char * argv [] )
...
     }
+
+
+  QStringList axesLabels = QStringList() << "W" << "E" << "S" << "N" << "Z" << "z";
+  if (!axes.setAxesLabels(axesLabels))
+    {
+    return EXIT_FAILURE;
+    }
+  QStringList axesLabelsControl = QStringList();
+  axesLabelsControl = axes.getAxesLabels();
```
Comment (jcfr) [inline Libs/Widgets/Testing/Cpp/ctkAxesWidgetTest1.cpp]: `getAxesLabels` -> `axesLabels()`

File: Libs/Widgets/Testing/Cpp/ctkAxesWidgetTest1.cpp [hunk 7]
```diff
@@ -84,6 +86,17 @@ int ctkAxesWidgetTest1(int argc, char * argv [] )
...
     return EXIT_FAILURE;
     }
+
+
+  QStringList axesLabels = QStringList() << "W" << "E" << "S" << "N" << "Z" << "z";
+  if (!axes.setAxesLabels(axesLabels))
+    {
+    return EXIT_FAILURE;
+    }
+  QStringList axesLabelsControl = QStringList();
```
Comment (jcfr) [inline Libs/Widgets/Testing/Cpp/ctkAxesWidgetTest1.cpp]: `QStringList axesLabelsControl;`

File: Libs/Widgets/Testing/Cpp/ctkAxesWidgetTest1.cpp [hunk 8]
```diff
@@ -84,6 +86,17 @@ int ctkAxesWidgetTest1(int argc, char * argv [] )
...
+  axesLabelsControl = axes.getAxesLabels();
+  CHECK_QSTRINGLIST(axesLabels, axesLabelsControl);
```
Comment (jcfr) [inline Libs/Widgets/Testing/Cpp/ctkAxesWidgetTest1.cpp]: `CHECK_QSTRINGLIST(axesLabels, axes.axesLabels())`

File: Libs/Widgets/ctkAxesWidget.cpp [hunk 9]
```diff
@@ -293,6 +293,27 @@ void ctkAxesWidget::paintEvent(QPaintEvent *)
...
+    qWarning("ctkAxesWidget::setAxesLabels() failed: Exactly 6 labels are expected.");
+    return false;
+    }
+
+  d->AxesLabels = labels;
+  return true;
+}
+
+// ----------------------------------------------------------------------------------
+QStringList ctkAxesWidget::getAxesLabels()
```
Comment (jcfr) [inline Libs/Widgets/ctkAxesWidget.cpp]: -> `QStringList ctkAxesWidget::axesLabels() const`

Add the `const`

File: Libs/Widgets/ctkAxesWidget.cpp [hunk 10]
```diff
@@ -293,6 +293,27 @@ void ctkAxesWidget::paintEvent(QPaintEvent *)
...
+{
+  Q_D(ctkAxesWidget);
```
Comment (jcfr) [inline Libs/Widgets/ctkAxesWidget.cpp]: `Q_D(const ctkAxesWidget);`

File: Libs/Widgets/ctkAxesWidget.h [hunk 11]
```diff
@@ -37,6 +37,7 @@ class CTK_WIDGETS_EXPORT ctkAxesWidget : public QWidget
   Q_ENUMS(Axis)
   Q_PROPERTY(Axis currentAxis READ currentAxis WRITE setCurrentAxis NOTIFY currentAxisChanged)
   Q_PROPERTY(bool autoReset READ autoReset WRITE setAutoReset)
+  Q_PROPERTY(QStringList AxesLabels READ getAxesLabels WRITE setAxesLabels)
```
Comment (jcfr) [inline Libs/Widgets/ctkAxesWidget.h]: `AxesLabels` -> `axesLabels`

File: Libs/Widgets/ctkAxesWidget.h [hunk 12]
```diff
@@ -81,6 +82,14 @@ public slots :
   /// Set the autoReset property to None anytime the currentAxis is changed.
   void setAutoReset(bool reset);
 
+  ///
+  /// Set the AxesLabels
+  bool setAxesLabels(const QStringList& labels);
+
+  ///
```
Comment (jcfr) [inline Libs/Widgets/ctkAxesWidget.h]: Remove extra `///`

File: Libs/Widgets/Testing/Cpp/ctkAxesWidgetTest1.cpp [hunk 13]
```diff
@@ -84,6 +86,17 @@ int ctkAxesWidgetTest1(int argc, char * argv [] )
             << spy[0].at(0).toInt() << " " << spy[1].at(0).toInt() << std::endl;
     return EXIT_FAILURE;
     }
+
+
+  QStringList axesLabels = QStringList() << "W" << "E" << "S" << "N" << "Z" << "z";
+  if (!axes.setAxesLabels(axesLabels))
```
Comment (jcfr) [inline Libs/Widgets/Testing/Cpp/ctkAxesWidgetTest1.cpp]: `CHECK_BOOL(axes.setAxesLabels(axesLabels), true)`

**Comment by Punzo** [comment]
@jcfr commits squashed

**Comment by jcfr** [comment]
Closing. This PR is superseded by #653

**Pull request closed without merging**